### PR TITLE
FirebaseAnimatedList client-side sorting, and other database plugin improvements

### PR DESCRIPTION
**Pull request by collinjackson**
This should address the database reading issues we've been seeing in the codelab/memechat, and expands API support a bit. Priority and sorting are out of scope for this PR, but coming soon.

Working on adding some tests, but would like reviews in the meantime.

### CODE REVIEW

**Comment by collinjackson** [comment]
@HansMuller can you please take a look at the AnimatedList usage and let me know if this is reasonable?

File: packages/firebase_database/lib/src/utils/push_id_generator.dart [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:math';
+
+/// Utility class for generating Firebase child node keys
+///
+/// Since the Flutter plugin API is asynchronous, there's no way for us
+/// to use the native SDK to generate node key synchronously and we
+/// have to do it ourselves if we want to be able to reference the
+/// newly-created node synchronously.
```
Comment (efortuna) [inline packages/firebase_database/lib/src/utils/push_id_generator.dart]: :-P

File: packages/firebase_database/lib/ui/firebase_animated_list.dart [hunk 2]
```diff
@@ -126,47 +128,61 @@ class FirebaseAnimatedList extends StatefulWidget {
...
   }
 
-  void _onValue() {
-    setState(() {
-      _dataAvailable = true;
-    });
-  }
-
   void _onChildAdded(int index, DataSnapshot snapshot) {
+    if (!_loaded)
```
Comment (efortuna) [inline packages/firebase_database/lib/ui/firebase_animated_list.dart]: nit: dart style says:
if (!_loaded) return;

File: packages/firebase_database/lib/ui/firebase_list.dart [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    onValue(event.snapshot);
+  }
+
+  int get length => _snapshots.length;
+  DataSnapshot operator [](int index) => _snapshots[index];
+  int indexOf(DataSnapshot item) => _snapshots.indexOf(item);
+
+  void dispose() {
+    assert(_subscriptions != null);
+    for (StreamSubscription<Event> subscription in _subscriptions)
```
Comment (efortuna) [inline packages/firebase_database/lib/ui/firebase_list.dart]: another nit: prefer {} around for loops, even single liners.

**Comment by HansMuller** [comment]
The new Flutter code LGTM.

File: packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java [hunk 4]
```diff
@@ -10,64 +12,152 @@
...
+  private FirebaseDatabasePlugin(MethodChannel channel) {
+    this.channel = channel;
+  }
+
+  private static DatabaseReference getReference(Map<String, ?> arguments) {
+    @SuppressWarnings("unchecked")
+    String path = (String) arguments.get("path");
+    DatabaseReference reference = FirebaseDatabase.getInstance().getReference();
+    if (path != null)
+      reference = reference.child(path);
```
Comment (collinjackson) [inline packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java]: Yeah I think Travis is going to make me dartfmt everything so unless I want to use curly braces I'll have to obey this convention.

**Comment by collinjackson** [comment]
The original change was split into two PRs:

* https://github.com/flutter/plugins/pull/62
* this one

Both were included in the 0.0.4 release.

**Pull request merged**